pkg/issuer/acme/dns/dnspod/dnspod.go

```diff
@@ -0,0 +1,153 @@
+// Package dnspod implements a DNS provider for solving the DNS-01 challenge
+// using dnspod DNS.
+package dnspod
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/decker502/dnspod-go"
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+)
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface.
+type DNSProvider struct {
+	client *dnspod.Client
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for dnspod.
+// Credentials must be passed in the environment variables: DNSPOD_API_KEY.
+func NewDNSProvider() (*DNSProvider, error) {
+	key := os.Getenv("DNSPOD_API_KEY")
+	return NewDNSProviderCredentials(key)
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for dnspod.
+func NewDNSProviderCredentials(key string) (*DNSProvider, error) {
+	if key == "" {
+		return nil, fmt.Errorf("dnspod credentials missing")
+	}
+
+	params := dnspod.CommonParams{LoginToken: key, Format: "json"}
+	return &DNSProvider{
+		client: dnspod.NewClient(params),
+	}, nil
+}
+
+// Present creates a TXT record to fulfil the dns-01 challenge.
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl := util.DNS01Record(domain, keyAuth)
+	zoneID, zoneName, err := c.getHostedZone(domain)
+	if err != nil {
+		return err
+	}
+
+	recordAttributes := c.newTxtRecord(zoneName, fqdn, value, ttl)
+	_, _, err = c.client.Domains.CreateRecord(zoneID, *recordAttributes)
```

Comment:
How does this behave if the record already exists? Present may be called multiple times, and it should not return an error if the record already exists already (it should update the value of the record instead)

Reply:
It's OK to call `Present` multiple times. If the record already exists, a new TXT record with the same name will be created, since a domain may have multiple TXT records associated with it. And all of them will be deleted when call `CleanUp`.

Reply:
This will result in a *lot* of duplicated entries though, which I don't think is what we want.

The Present function may be called ~10 times (in the best case scenario) right now.

Does the DNSPod API support 'create or update' operations, like other providers?

Reply:
If the value is same as before, then there will be no duplicated entries, the server will return OK without any changes. For example, create a TXT record `_acme-challenge.test` with value `sometext`, then send same request multiple times, there will be only one TXT record.
if send another request to create same TXT record `_acme-challenge.test` with value `some othertext`, there will be two TXT records.

The DNSPod API does not support 'create or update' operations, and if update operation be called multiple times without any changes, the record will be locked for 1 hour.